### DISCUSSION TITLE
Mod crashes the game "java.lang.RuntimeException: Could not execute entrypoint stage 'client' due to errors, provided by 'collar-fabric'!"

### DISCUSSION THREAD

**Discussion by CaptainRexPL** [FabricMC/fabric-api#2602] [Mod Dev Support] [answered]
I updated the mod from 1.17 to 1.19.2 and fixed all the errors (at least I think that everything should work fine). When I try to launch the game with this mod I get this message:
`---- Minecraft Crash Report ----
// There are four lights!

Time: 2022-10-22 14:05:55
Description: Initializing game

java.lang.RuntimeException: Could not execute entrypoint stage 'client' due to errors, provided by 'collar-fabric'!
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.lambda$invoke0$0(EntrypointUtils.java:51)
	at net.fabricmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:49)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke(EntrypointUtils.java:35)
	at net.fabricmc.loader.impl.game.minecraft.Hooks.startClient(Hooks.java:53)
	at net.minecraft.class_310.<init>(class_310.java:459)
	at net.minecraft.client.main.Main.method_44604(Main.java:205)
	at net.minecraft.client.main.Main.main(Main.java:51)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:461)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
Caused by: java.lang.NullPointerException: Cannot invoke "com.mojang.brigadier.CommandDispatcher.register(com.mojang.brigadier.builder.LiteralArgumentBuilder)" because "dispatcher" is null
	at com.collarmc.mod.common.commands.Commands.registerServiceCommands(Commands.java:81)
	at com.collarmc.mod.common.commands.Commands.register(Commands.java:58)
	at com.collarmc.mod.fabric.client.CollarFabricClient.onInitializeClient(CollarFabricClient.java:36)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:47)
	... 8 more

A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Render thread
Stacktrace:
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.lambda$invoke0$0(EntrypointUtils.java:51)
	at net.fabricmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:49)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke(EntrypointUtils.java:35)
	at net.fabricmc.loader.impl.game.minecraft.Hooks.startClient(Hooks.java:53)
	at net.minecraft.class_310.<init>(class_310.java:459)Stacktrace:
	at net.minecraft.client.main.Main.method_44604(Main.java:205)
	at net.minecraft.client.main.Main.main(Main.java:51)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:461)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)

-- System Details --
Details:
	Minecraft Version: 1.19.2
	Minecraft Version ID: 1.19.2
	Operating System: Windows 10 (amd64) version 10.0
	Java Version: 17.0.3, Microsoft
	Java VM Version: OpenJDK 64-Bit Server VM (mixed mode), Microsoft
	Memory: 950321448 bytes (906 MiB) / 2483027968 bytes (2368 MiB) up to 3221225472 bytes (3072 MiB)
	CPUs: 6
	Processor Vendor: AuthenticAMD
	Processor Name: AMD Ryzen 5 4500U with Radeon Graphics         
	Identifier: AuthenticAMD Family 23 Model 96 Stepping 1
	Microarchitecture: unknown
	Frequency (GHz): 2.37
	Number of physical packages: 1
	Number of physical CPUs: 6
	Number of logical CPUs: 6
	Graphics card #0 name: AMD Radeon(TM) Graphics
	Graphics card #0 vendor: Advanced Micro Devices, Inc. (0x1002)
	Graphics card #0 VRAM (MB): 512.00
	Graphics card #0 deviceId: 0x1636
	Graphics card #0 versionInfo: DriverVersion=26.20.14040.0
	Memory slot #0 capacity (MB): 8192.00
	Memory slot #0 clockSpeed (GHz): 2.67
	Memory slot #0 type: DDR4
	Memory slot #1 capacity (MB): 8192.00
	Memory slot #1 clockSpeed (GHz): 2.67
	Memory slot #1 type: DDR4
	Virtual memory max (MB): 22463.00
	Virtual memory used (MB): 19521.07
	Swap memory total (MB): 6721.70
	Swap memory used (MB): 1168.63
	JVM Flags: 9 total; -XX:HeapDumpPath=MojangTricksIntelDriversForPerformance_javaw.exe_minecraft.exe.heapdump -Xss1M -Xmx3G -XX:+UnlockExperimentalVMOptions -XX:+UseG1GC -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=32M
	Fabric Mods: 
		clientcommands: Client Commands 2.7.3
			clientarguments: clientarguments 1.4.1
			multiconnect-api: MultiConnect API 1.5.10
		collar-fabric: Collar Fabric 0.1-397b605
		fabric-api: Fabric API 0.60.0+1.19.2
			fabric-api-base: Fabric API Base 0.4.11+e62f51a390
			fabric-api-lookup-api-v1: Fabric API Lookup API (v1) 1.6.9+9ff28f4090
			fabric-biome-api-v1: Fabric Biome API (v1) 9.0.17+aeb40ebe90
			fabric-blockrenderlayer-v1: Fabric BlockRenderLayer Registration (v1) 1.1.20+aeb40ebe90
			fabric-client-tags-api-v1: Fabric Client Tags 1.0.1+b35fea8390
			fabric-command-api-v1: Fabric Command API (v1) 1.2.11+f71b366f90
			fabric-command-api-v2: Fabric Command API (v2) 2.1.7+0c17ea9690
			fabric-commands-v0: Fabric Commands (v0) 0.2.28+df3654b390
			fabric-containers-v0: Fabric Containers (v0) 0.1.34+df3654b390
			fabric-content-registries-v0: Fabric Content Registries (v0) 3.2.3+aeb40ebe90
			fabric-convention-tags-v1: Fabric Convention Tags 1.1.1+7cd20a1490
			fabric-crash-report-info-v1: Fabric Crash Report Info (v1) 0.2.5+aeb40ebe90
			fabric-data-generation-api-v1: Fabric Data Generation API (v1) 5.1.8+a680b9b490
			fabric-dimensions-v1: Fabric Dimensions API (v1) 2.1.30+aeb40ebe90
			fabric-entity-events-v1: Fabric Entity Events (v1) 1.4.18+9ff28f4090
			fabric-events-interaction-v0: Fabric Events Interaction (v0) 0.4.28+aeb40ebe90
			fabric-events-lifecycle-v0: Fabric Events Lifecycle (v0) 0.2.28+df3654b390
			fabric-game-rule-api-v1: Fabric Game Rule API (v1) 1.0.21+aeb40ebe90
			fabric-item-api-v1: Fabric Item API (v1) 1.5.7+35a03c4390
			fabric-item-groups-v0: Fabric Item Groups (v0) 0.3.28+aeb40ebe90
			fabric-key-binding-api-v1: Fabric Key Binding API (v1) 1.0.20+aeb40ebe90
			fabric-keybindings-v0: Fabric Key Bindings (v0) 0.2.18+df3654b390
			fabric-lifecycle-events-v1: Fabric Lifecycle Events (v1) 2.1.2+aeb40ebe90
			fabric-loot-api-v2: Fabric Loot API (v2) 1.1.3+83a8659290
			fabric-loot-tables-v1: Fabric Loot Tables (v1) 1.1.6+9e7660c690
			fabric-message-api-v1: Fabric Message API (v1) 5.0.3+176380a290
			fabric-mining-level-api-v1: Fabric Mining Level API (v1) 2.1.14+33fbc73890
			fabric-models-v0: Fabric Models (v0) 0.3.17+aeb40ebe90
			fabric-networking-api-v1: Fabric Networking API (v1) 1.2.4+5eb68ef290
			fabric-networking-v0: Fabric Networking (v0) 0.3.21+df3654b390
			fabric-object-builder-api-v1: Fabric Object Builder API (v1) 4.0.11+aeb40ebe90
			fabric-particles-v1: Fabric Particles (v1) 1.0.10+aeb40ebe90
			fabric-registry-sync-v0: Fabric Registry Sync (v0) 0.9.24+aeb40ebe90
			fabric-renderer-api-v1: Fabric Renderer API (v1) 1.0.10+aeb40ebe90
			fabric-renderer-indigo: Fabric Renderer - Indigo 0.6.12+aeb40ebe90
			fabric-renderer-registries-v1: Fabric Renderer Registries (v1) 3.2.20+df3654b390
			fabric-rendering-data-attachment-v1: Fabric Rendering Data Attachment (v1) 0.3.14+aeb40ebe90
			fabric-rendering-fluids-v1: Fabric Rendering Fluids (v1) 3.0.7+aeb40ebe90
			fabric-rendering-v0: Fabric Rendering (v0) 1.1.22+df3654b390
			fabric-rendering-v1: Fabric Rendering (v1) 1.10.16+aeb40ebe90
			fabric-resource-conditions-api-v1: Fabric Resource Conditions API (v1) 2.0.11+e62f51a390
			fabric-resource-loader-v0: Fabric Resource Loader (v0) 0.6.1+aeb40ebe90
			fabric-screen-api-v1: Fabric Screen API (v1) 1.0.26+f05effaf90
			fabric-screen-handler-api-v1: Fabric Screen Handler API (v1) 1.2.13+9ff28f4090
			fabric-textures-v0: Fabric Textures (v0) 1.0.20+aeb40ebe90
			fabric-transfer-api-v1: Fabric Transfer API (v1) 2.0.11+83a8659290
			fabric-transitive-access-wideners-v1: Fabric Transitive Access Wideners (v1) 1.3.0+42d99c3290
		fabricloader: Fabric Loader 0.14.9
		java: OpenJDK 64-Bit Server VM 17		
		minecraft: Minecraft 1.19.2
		modmenu: Mod Menu 4.0.6
		nochatreports: No Chat Reports 1.19.2-v1.13.9
	        sodium: Sodium 0.4.2+build.16
			org_joml_joml: joml 1.10.4
		viabackwards: ViaBackwards 4.4.2
		viafabric: ViaFabric 0.4.9+19-main
			org_yaml_snakeyaml: snakeyaml 1.31
			viafabric-mc119: ViaFabric for 1.19 0.4.9+19-main
		viarewind: ViaRewind 2.0.2
		viaversion: ViaVersion 4.4.2
		xaerominimap: Xaero's Minimap 22.13.2
	Launched Version: fabric-loader-0.14.9-1.19.2
	Backend library: LWJGL version 3.3.1 SNAPSHOT
	Backend API: Unknown
	Window size: <not initialized>
	GL Caps: Using framebuffer using OpenGL 3.2
	GL debug messages: <disabled>
	Using VBOs: Yes
	Is Modded: Definitely; Client brand changed to 'fabric'
	Type: Client (map_client.txt)
	CPU: <unknown>` 
	
	minecraft version: 1.19.2
	fabric api version: 0.60.0
	brigadier version: 1.0.18
	fabric version: 1.14.9
	
	CollarFabricClient class:
`package com.collarmc.mod.fabric.client;

import com.collarmc.client.plugin.Plugins;
import com.collarmc.mod.common.CollarService;
import com.collarmc.mod.common.commands.Commands;
import com.collarmc.mod.common.events.CollarModInitializedEvent;
import com.collarmc.mod.common.features.messaging.Messages;
import com.collarmc.mod.common.plastic.CollarTextureProvider;
import com.collarmc.mod.glue.render.TracerRenderer;
import com.collarmc.mod.glue.render.WaypointRenderer;
import com.collarmc.plastic.Plastic;
import com.collarmc.plastic.FabricPlastic;
import com.collarmc.pounce.EventBus;
import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.command.v2.ClientCommandManager;
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource;

import java.io.IOException;

@Environment(EnvType.CLIENT)
public class CollarFabricClient implements ClientModInitializer {

    private static final EventBus EVENT_BUS = new EventBus(Runnable::run);
    private static final Plugins PLUGINS = new Plugins();
    private static final Plastic PLASTIC = new FabricPlastic(new CollarTextureProvider(), EVENT_BUS);
    private static final CollarService COLLAR_SERVICE = new CollarService(PLASTIC, EVENT_BUS, PLUGINS);
    private static final WaypointRenderer WAYPOINT_RENDERER = new WaypointRenderer(PLASTIC, COLLAR_SERVICE);
    private static final TracerRenderer TRACER_RENDERER = new TracerRenderer(PLASTIC, COLLAR_SERVICE);
    private static final Messages GROUP_CHAT_SERVICE = new Messages(PLASTIC, COLLAR_SERVICE);

    @Override
    public void onInitializeClient() {
        Commands<FabricClientCommandSource> commands = new Commands<>(COLLAR_SERVICE, GROUP_CHAT_SERVICE, PLASTIC, true);
        commands.register(ClientCommandManager.getActiveDispatcher());
        EVENT_BUS.subscribe(WAYPOINT_RENDERER);
        EVENT_BUS.subscribe(TRACER_RENDERER);
        EVENT_BUS.dispatch(new CollarModInitializedEvent());
        try {
            PLUGINS.loadPlugins(CollarFabricClient.class.getClassLoader(), EVENT_BUS);
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }
}`

Commands class:
`package com.collarmc.mod.common.commands;

import com.collarmc.api.friends.Friend;
import com.collarmc.api.friends.Status;
import com.collarmc.api.groups.Group;
import com.collarmc.api.groups.GroupType;
import com.collarmc.api.location.Dimension;
import com.collarmc.api.location.Location;
import com.collarmc.api.minecraft.MinecraftPlayer;
import com.collarmc.api.waypoints.Waypoint;
import com.collarmc.client.api.groups.GroupInvitation;
import com.collarmc.mod.common.CollarService;
import com.collarmc.mod.common.commands.arguments.*;
import com.collarmc.mod.common.commands.arguments.IdentityArgumentType.IdentityArgument;
import com.collarmc.mod.common.commands.arguments.WaypointArgumentType.WaypointArgument;
import com.collarmc.mod.common.features.messaging.Messages;
import com.collarmc.plastic.Plastic;
import com.collarmc.plastic.player.Player;
import com.collarmc.plastic.ui.TextColor;
import com.google.common.collect.ImmutableList;
import com.mojang.brigadier.Command;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.builder.ArgumentBuilder;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import static com.collarmc.mod.common.commands.arguments.DimensionArgumentType.dimension;
import static com.collarmc.mod.common.commands.arguments.GroupArgumentType.getGroup;
import static com.collarmc.mod.common.commands.arguments.InvitationArgumentType.getInvitation;
import static com.collarmc.mod.common.commands.arguments.PlayerArgumentType.getPlayer;
import static com.mojang.brigadier.arguments.DoubleArgumentType.doubleArg;
import static com.mojang.brigadier.arguments.DoubleArgumentType.getDouble;
import static com.mojang.brigadier.arguments.StringArgumentType.getString;
import static com.mojang.brigadier.arguments.StringArgumentType.string;

public final class Commands<S> {

    private final CollarService collarService;
    private final Messages messages;
    private final Plastic plastic;
    private final boolean prefixed;

    public Commands(CollarService collarService, Messages messages, Plastic plastic, boolean prefixed) {
        this.collarService = collarService;
        this.messages = messages;
        this.plastic = plastic;
        this.prefixed = prefixed;
    }

    public void register(CommandDispatcher<S> dispatcher) {
        registerServiceCommands(dispatcher);
        registerFriendCommands(dispatcher);
        registerLocationCommands(dispatcher);
        registerWaypointCommands(dispatcher);
        registerGroupCommands(GroupType.PARTY, dispatcher);
        registerGroupCommands(GroupType.GROUP, dispatcher);
        registerChatCommands(dispatcher);
    }

    private LiteralArgumentBuilder<S> prefixed(String name, LiteralArgumentBuilder<S> argumentBuilder) {
        return this.prefixed ? literal("collar").then(literal(name).then(argumentBuilder)) : literal(name).then(argumentBuilder);
    }

    private LiteralArgumentBuilder<S> prefixed(String name, ArgumentBuilder<S, ?> argument) {
        return this.prefixed ? literal("collar").then(literal(name).then(argument)) : literal(name).then(argument);
    }

    private LiteralArgumentBuilder<S> prefixed(String name, Command<S> command) {
        return this.prefixed ? literal("collar").then(literal(name).executes(command)) : literal(name).executes(command);
    }

    private void registerServiceCommands(CommandDispatcher<S> dispatcher) {
        // collar connect
        dispatcher.register(prefixed("connect", context -> {
            collarService.connect();
            return 1;
        }));

        // collar disconnect
        dispatcher.register(prefixed("disconnect", context -> {
            collarService.disconnect();
            return 1;
        }));

        // collar status
        dispatcher.register(prefixed("status", context -> {
            collarService.with(collar -> {
                plastic.display.displayInfoMessage("Collar is " + collar.getState().name().toLowerCase());
            }, () -> plastic.display.displayMessage("Collar is disconnected"));
            return 1;
        }));

        // collar me
        dispatcher.register(prefixed("me", context -> {
            collarService.with(collar -> {
                collar.identities().resolveProfile(collar.player()).thenAccept(publicProfile -> {
                    Player player = plastic.world.findPlayerById(collar.player().minecraftPlayer.id).orElseThrow(() -> new IllegalStateException("should have been able to find self"));
                    publicProfile.ifPresent(profile -> plastic.display.displayInfoMessage("You are connected as " + profile.name + " on minecraft account " + player.name()));
                });
            });
            return 1;
        }));
    }

    private void registerFriendCommands(CommandDispatcher<S> dispatcher) {
        // collar friend add [user]
        dispatcher.register(prefixed("friend", literal("add")
                .then(argument("name", identity())
                        .executes(context -> {
                            collarService.with(collar -> {
                                IdentityArgument player = context.getArgument("name", IdentityArgument.class);
                                if (player.player != null) {
                                    collar.friends().addFriend(new MinecraftPlayer(player.player.id(), collar.player().minecraftPlayer.server, collar.player().minecraftPlayer.networkId));
                                } else if (player.profile != null) {
                                    collar.friends().addFriend(player.profile.id);
                                }
                            });
                            return 1;
                        }))));

        // collar friend remove [user]
        dispatcher.register(prefixed("friend", literal("remove")
                .then(argument("name", identity())
                        .executes(context -> {
                            collarService.with(collar -> {
                                IdentityArgument player = context.getArgument("name", IdentityArgument.class);
                                if (player.player != null) {
                                    collar.friends().removeFriend(new MinecraftPlayer(player.player.id(), collar.player().minecraftPlayer.server, collar.player().minecraftPlayer.networkId));
                                } else if (player.profile != null) {
                                    collar.friends().removeFriend(player.profile.id);
                                } else {
                                    throw new IllegalStateException("was not profile or player");
                                }
                            });
                            return 1;
                        }))));

        // collar friend list
        dispatcher.register(prefixed("friend", literal("list")
                .executes(context -> {
                    collarService.with(collar -> {
                        Set<Friend> friends = collar.friends().list();
                        if (friends.isEmpty()) {
                            plastic.display.displayInfoMessage("You don't have any friends");
                        } else {
                            friends.stream().sorted(Comparator.comparing(o -> o.status)).forEach(friend -> {
                                TextColor color = friend.status.equals(Status.ONLINE) ? TextColor.GREEN : TextColor.GRAY;
                                plastic.display.displayMessage(plastic.display.newTextBuilder().add(friend.profile.name, color));
                            });
                        }
                    });
                    return 1;
                })));
    }

    private void registerGroupCommands(GroupType type, CommandDispatcher<S> dispatcher) {
        // collar party create [name]
        dispatcher.register(prefixed(type.name, literal("create")
                .then(argument("name", string())
                        .executes(context -> {
                            collarService.with(collar -> {
                                collar.groups().create(getString(context, "name"), type, ImmutableList.of());
                            });
                            return 1;
                        }))));

        // collar party delete [name]
        dispatcher.register(prefixed(type.name, literal("delete")
                .then(argument("name", group(type))
                        .executes(context -> {
                            collarService.with(collar -> {
                                collar.groups().delete(getGroup(context, "name"));
                            });
                            return 1;
                        }))));

        // collar party leave [name]
        dispatcher.register(prefixed(type.name, literal("leave")
                .then(argument("name", group(type))
                        .executes(context -> {
                            collarService.with(collar -> {
                                collar.groups().leave(getGroup(context, "name"));
                            });
                            return 1;
                        }))));

        // collar par/**/ty invites
        dispatcher.register(prefixed(type.name, literal("invites")
                .executes(context -> {
                    collarService.with(collar -> {
                        List<GroupInvitation> invitations = collar.groups().invitations().stream()
                                .filter(invitation -> invitation.type == type)
                                .collect(Collectors.toList());
                        if (invitations.isEmpty()) {
                            plastic.display.displayInfoMessage("You have no invites to any " + type.plural);
                        } else {
                            plastic.display.displayInfoMessage("You have invites to:");
                            invitations.forEach(invitation -> plastic.display.displayInfoMessage(invitation.name));
                            plastic.display.displayInfoMessage("To accept type '/collar " + type.name  + " accept [name]");
                        }
                    });
                    return 1;
                })));

        // collar party accept [name]
        dispatcher.register(prefixed(type.name, literal("accept")
                .then(argument("groupName", invitation(type))
                        .executes(context -> {
                            collarService.with(collar -> {
                                collar.groups().accept(getInvitation(context, "groupName"));
                            });
                            return 1;
                        }))));

        // collar party list
        dispatcher.register(prefixed(type.name, literal("list")
                .executes(context -> {
                    collarService.with(collar -> {
                        List<Group> parties = collar.groups().all().stream()
                                .filter(group -> group.type.equals(type))
                                .collect(Collectors.toList());
                        if (parties.isEmpty()) {
                            plastic.display.displayInfoMessage("You are not a member of any " + type.plural);
                        } else {
                            plastic.display.displayInfoMessage("You belong to the following " + type.plural + ":");
                            parties.forEach(group -> plastic.display.displayInfoMessage(group.name));
                        }
                    });
                    return 1;
                })));

        // collar party add [name] [player]
        dispatcher.register(prefixed(type.name, literal("add")
                .then(argument("groupName", group(type))
                        .then(argument("playerName", player())
                                .executes(context -> {
                                    collarService.with(collar -> {
                                        Group group = getGroup(context, "groupName");
                                        Player player = getPlayer(context, "playerName");
                                        collar.groups().invite(group, ImmutableList.of(player.id()));
                                    });
                                    return 1;
                                })))));

        // collar party remove [name] [player]
        dispatcher.register(prefixed(type.name, literal("remove")
                .then(argument("groupName", group(type))
                        .then(argument("playerName", identity())
                                .executes(context -> {
                                    collarService.with(collar -> {
                                        Group group = getGroup(context, "groupName");
                                        IdentityArgument identity = context.getArgument("playerName", IdentityArgument.class);
                                        group.members.stream().filter(candidate -> candidate.profile.id.equals(identity.profile.id)).findFirst().ifPresent(theMember -> {
                                            collar.groups().removeMember(group, theMember);
                                        });
                                    });
                                    return 1;
                                })))));

        // collar party members [name]
        dispatcher.register(prefixed(type.name, literal("members")
                .then(argument("groupName", group(type)))
                .executes(context -> {
                    collarService.with(collar -> {
                        Group group = getGroup(context, "groupName");
                        plastic.display.displayMessage("Members:");
                        group.members.forEach(member -> {
                            Optional<Player> thePlayer = plastic.world.allPlayers().stream().filter(player -> member.player.minecraftPlayer.id.equals(player.id())).findFirst();
                            String message;
                            if (thePlayer.isPresent()) {
                                message = member.profile.name + " playing as " + member.player.minecraftPlayer.id + " (" + member.membershipRole.name() + ")";
                            } else {
                                message = member.profile.name;
                            }
                            plastic.display.displayMessage(message + "(" + member.membershipRole.name() + ")");
                        });
                    });
                    return 1;
                })));
    }

    private void registerLocationCommands(CommandDispatcher<S> dispatcher) {
        // collar location share start [any group name]
        dispatcher.register(prefixed("location", literal("share")
                .then(literal("start")
                        .then(argument("groupName", groups())
                                .executes(context -> {
                                    collarService.with(collar -> {
                                        Group group = getGroup(context, "groupName");
                                        collar.location().startSharingWith(group);
                                    });
                                    return 1;
                                })))));

        // collar location share stop [any group name]
        dispatcher.register(prefixed("location", literal("share")
                .then(literal("stop")
                        .then(argument("groupName", groups())
                                .executes(context -> {
                                    collarService.with(collar -> {
                                        Group group = getGroup(context, "groupName");
                                        collar.location().stopSharingWith(group);
                                    });
                                    return 1;
                                })))));

        // collar location share stop [any group name]
        dispatcher.register(prefixed("location", literal("share")
                .then(literal("list")
                        .executes(context -> {
                            collarService.with(collar -> {
                                List<Group> active = collar.groups().all().stream()
                                        .filter(group -> collar.location().isSharingWith(group))
                                        .collect(Collectors.toList());
                                if (active.isEmpty()) {
                                    plastic.display.displayInfoMessage("You are not sharing your location with any groups");
                                } else {
                                    plastic.display.displayInfoMessage("You are sharing your location with groups:");
                                    active.forEach(group -> plastic.display.displayInfoMessage(group.name + " (" + group.type.name + ")"));
                                }
                            });
                            return 1;
                        }))));
    }

    private void registerWaypointCommands(CommandDispatcher<S> dispatcher) {

        // collar location waypoint add [name]
        dispatcher.register(prefixed("waypoint", literal("add")
                .then(argument("name", string())
                        .executes(context -> {
                            collarService.with(collar -> {
                                Location location = plastic.world.currentPlayer().location();
                                collar.location().addWaypoint(getString(context, "name"), location);
                            });
                            return 1;
                        }))));

        // collar waypoint remove [name]
        dispatcher.register(prefixed("waypoint", literal("remove")
                .then(argument("name", privateWaypoint())
                        .executes(context -> {
                            collarService.with(collar -> {
                                WaypointArgument argument = context.getArgument("name", WaypointArgument.class);
                                collar.location().removeWaypoint(argument.waypoint);
                            });
                            return 1;
                        }))));

        // collar location waypoint list
        dispatcher.register(prefixed("waypoint", literal("list")
                .executes(context -> {
                    collarService.with(collar -> {
                        Set<Waypoint> waypoints = collar.location().privateWaypoints();
                        if (waypoints.isEmpty()) {
                            plastic.display.displayInfoMessage("You have no private waypoints");
                        } else {
                            waypoints.forEach(waypoint -> plastic.display.displayInfoMessage(waypoint.displayName()));
                        }
                    });
                    return 1;
                })));

        // collar location waypoint list [any group name]
        dispatcher.register(prefixed("waypoint", literal("list")
                .then(argument("group", groups())
                        .executes(context -> {
                            collarService.with(collar -> {
                                Group group = getGroup(context, "group");
                                Set<Waypoint> waypoints = collar.location().groupWaypoints(group);
                                if (waypoints.isEmpty()) {
                                    plastic.display.displayInfoMessage("You have no group waypoints");
                                } else {
                                    waypoints.forEach(waypoint -> plastic.display.displayInfoMessage(waypoint.displayName()));
                                }
                            });
                            return 1;
                        }))));

        // collar location waypoint add [name] [x] [y] [z] to [group]
        dispatcher.register(prefixed("waypoint", literal("add")
                .then(argument("name", string())
                        .then(argument("x", doubleArg())
                                .then(argument("y", doubleArg())
                                        .then(argument("z", doubleArg())
                                                .then(argument("dimension", dimension())
                                                        .then(literal("to")
                                                                .then(argument("group", groups())
                                                                        .executes(context -> {
                                                                            collarService.with(collar -> {
                                                                                Group group = getGroup(context, "group");
                                                                                Dimension dimension = context.getArgument("dimension", Dimension.class);
                                                                                Location location = new Location(
                                                                                        getDouble(context, "x"),
                                                                                        getDouble(context, "y"),
                                                                                        getDouble(context, "z"),
                                                                                        dimension
                                                                                );
                                                                                collar.location().addWaypoint(group, getString(context, "name"), location);
                                                                            });
                                                                            return 1;
                                                                        }))))))))));

        // collar location waypoint remove [name] from [group]
        dispatcher.register(prefixed("waypoint", literal("remove")
                .then(argument("name", groupWaypoint())
                        .then(literal("from")
                                .then(argument("group", groups())
                                        .executes(context -> {
                                            collarService.with(collar -> {
                                                Group group = getGroup(context, "group");
                                                WaypointArgument argument = context.getArgument("waypoint", WaypointArgument.class);
                                                if (!group.id.equals(argument.group.id)) {
                                                    collar.location().removeWaypoint(argument.group, argument.waypoint);
                                                } else {
                                                    plastic.display.displayInfoMessage("Waypoint " + argument.waypoint + " does not belong to group " + group.name);
                                                }
                                            });
                                            return 1;
                                        }))))));

        // collar waypoint add [name] [x] [y] [z]
        dispatcher.register(prefixed("waypoint", literal("add")
                .then(argument("name", string())
                        .then(argument("x", doubleArg())
                                .then(argument("y", doubleArg())
                                        .then(argument("z", doubleArg())
                                                .then(argument("dimension", dimension())
                                                        .executes(context -> {
                                                            collarService.with(collar -> {
                                                                Dimension dimension = context.getArgument("dimension", Dimension.class);
                                                                Location location = new Location(
                                                                        getDouble(context, "x"),
                                                                        getDouble(context, "y"),
                                                                        getDouble(context, "z"),
                                                                        dimension
                                                                );
                                                                collar.location().addWaypoint(getString(context, "name"), location);
                                                            });
                                                            return 1;
                                                        }))))))));
    }

    private void registerChatCommands(CommandDispatcher<S> dispatcher) {
        // /msg player2 OwO
        dispatcher.register(literal("msg")
                .then(argument("recipient", player())
                        .then(argument("rawMessage", string())
                                .executes(context -> {
                                    Player recipient = getPlayer(context, "recipient");
                                    String message = getString(context, "rawMessage");
                                    messages.sendMessage(recipient, message);
                                    return 1;
                                }))));

        // collar chat with coolkids
        dispatcher.register(prefixed("chat", literal("with")
                .then(argument("group", groups())
                        .executes(context -> {
                            Group group = getGroup(context, "group");
                            messages.switchToGroup(group);
                            return 1;
                        }))));

        // collar chat off
        dispatcher.register(prefixed("chat", literal("off").executes(context -> {
            messages.switchToGeneralChat();
            return 1;
        })));
    }

    public <T> RequiredArgumentBuilder<S, T> argument(String name, ArgumentType<T> type) {
        return RequiredArgumentBuilder.argument(name, type);
    }

    private LiteralArgumentBuilder<S> literal(final String name) {
        return LiteralArgumentBuilder.literal(name);
    }

    private GroupArgumentType group(GroupType type) {
        return new GroupArgumentType(collarService, type);
    }

    private GroupArgumentType groups() {
        return new GroupArgumentType(collarService, null);
    }

    private InvitationArgumentType invitation(GroupType type) {
        return new InvitationArgumentType(collarService, type);
    }

    private WaypointArgumentType privateWaypoint() {
        return new WaypointArgumentType(collarService, true);
    }

    private WaypointArgumentType groupWaypoint() {
        return new WaypointArgumentType(collarService, false);
    }

    private PlayerArgumentType player() {
        return new PlayerArgumentType(plastic);
    }

    private IdentityArgumentType identity() {
        return new IdentityArgumentType(collarService, plastic);
    }

    private GroupMemberArgumentType groupMember() {
        return new GroupMemberArgumentType(collarService, plastic);
    }
}`

**Comment by apple502j** [comment]
You should use Command API v2; see https://fabricmc.net/2022/06/07/119.html

**Reply by CaptainRexPL** [reply]
I checked and all dependencies in this mod and they are all up-to-date. The only mod that I has v1 version is fabric API according to the game crash log.

**Comment by warjort** [comment, marked as answer]
You have: ````fabric-command-api-v2: Fabric Command API (v2) 2.1.7+0c17ea9690````

See: https://github.com/FabricMC/fabric/blob/1.19.2/fabric-command-api-v2/src/client/java/net/fabricmc/fabric/api/client/command/v2/ClientCommandRegistrationCallback.java
